File: backend/app/schemas/mcp.py
"""What an MCP client needs to reach this library, and what it gets."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class McpTool(BaseModel):
    name: str
    #: The tool's own docstring, dedented. Its first sentence is the blurb.
    description: str


class McpRecipe(BaseModel):
    """The connection recipe — a command, because the transport is stdio.

    The client spawns ``command args`` itself; there is no socket to report.
    ``env`` is empty unless the API is off its default port, in which case it
    pins ``SCINET_API_URL`` so the spawned server finds it.
    """

    transport: Literal["stdio"] = "stdio"
    name: str
    command: str
    args: list[str]
    backend_dir: str
    api_url: str
    env: dict[str, str]
    available: bool
    tools: list[McpTool]

File: frontend/src/api/mcp.ts
/**
 * The MCP connection recipe, from the API.
 *
 * The transport is stdio, so there is nothing to connect to: the recipe is
 * the command an MCP client spawns, with this checkout's path filled in, and
 * the tools the server reports about itself. `env` is empty unless the API
 * is off its default port.
 */

export interface McpTool {
  name: string
  /** The tool's own docstring, dedented; the first sentence is the blurb. */
  description: string
}

export interface McpRecipe {
  transport: 'stdio'
  /** The key under `mcpServers` — "scinet". */
  name: string
  command: string
  args: string[]
  backend_dir: string
  api_url: string
  env: Record<string, string>
  /** False only if the mcp package will not import in backend/.venv. */
  available: boolean
  tools: McpTool[]
}

export async function fetchMcpRecipe(): Promise<McpRecipe> {
  const res = await fetch('/api/mcp')
  if (!res.ok) throw new Error(`/api/mcp -> ${res.status}`)
  return (await res.json()) as McpRecipe
}
